common/src/reserved.rs

```diff
@@ -59,18 +62,72 @@ impl ReservedState {
                     .or_insert(member.consensus_voting_power);
             }
         }
-        Ok(governance_set
-            .iter()
-            .map(|(name, voting_power)| (self.query_public_key(name).unwrap(), *voting_power))
-            .collect())
+        let mut result = Vec::new();
+        for (name, voting_power) in governance_set {
+            let public_key = self.query_public_key(&name).ok_or_else(|| {
+                format!("The public key of {name} is not found in the genesis info.")
+            })?;
+            result.push((public_key, voting_power));
+        }
+        Ok(result)
     }
 
-    pub fn apply_delegate(&mut self, _tx: &TxDelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_delegate(&mut self, tx: &TxDelegate) -> Result<Self, String> {
+        let data = DelegationTransactionData {
+            delegator: tx.delegator.clone(),
+            delegatee: tx.delegatee.clone(),
+            governance: tx.governance,
+            timestamp: tx.timestamp as BlockHeight,
+        };
+        match tx.proof.verify(&data) {
+            Ok(_) => true,
+            Err(_) => return Err("proof verification failed".to_string()),
```

Comment:
```suggestion
            Err(_) => return Err("delegation proof verification failed".to_string()),
```
Providing enough context in an error message I think is a better practice.

Reply:
would change it